File: EmBeLife/Models/GiftModels.swift
import Foundation

/// Navigation steps for gift giver (1–4) and receiver (5) experiences.
enum GiftExperienceRoute: Hashable {
    case confirm
    case paymentMethod
    case sent
    case signUp
    case received
}

/// Draft state for the Gift $ Amount flow.
@Observable
final class GiftDraft {
    /// Preset gift amounts user can pick without typing.
    static let amountPresets: [Double] = [50, 100, 200, 500, 1_000, 2_000]

    /// Backing typed amount for the number pad field (digits only).
    var amountInput: String = "1000"
    var selectedMethod: PaymentMethodKind = .creditCard
    var bankDetails = BankAccountDetails.sample
    var zelleDetails = ContactPaymentDetails.zelleSample
    var venmoDetails = ContactPaymentDetails.venmoSample
    var paypalDetails = ContactPaymentDetails.paypalSample
    var creditCardDetails = CreditCardDetails.sample

    /// Display recipient for giver confirm screen.
    var recipientLabel = "Gift money to EmBeLife User"
    /// Sample receiver attribution for experience_5.
    var senderDisplayName = "J. R"
    var receivedDate = Calendar.current.date(from: DateComponents(year: 2026, month: 5, day: 27))
        ?? Date()

    /// Payment methods available when funding a gift (excludes Gift Fund).
    static let giftPaymentMethods: [PaymentMethodKind] = [
        .bankAccount, .zelle, .venmo, .paypal, .creditCard
    ]

    var amount: Double {
        get {
            let cleaned = amountInput.filter { $0.isNumber || $0 == "." }
            return Double(cleaned) ?? 0
        }
        set {
            let value = max(0, newValue)
            if value == floor(value) {
                amountInput = String(Int(value))
            } else {
                amountInput = String(format: "%.2f", value)
            }
        }
    }

    var isValidGiftAmount: Bool {
        amount > 0
    }

    var amountLabel: String {
        Self.formatCurrency(amount)
    }

    func applyPreset(_ preset: Double) {
        amount = preset
    }

    func sanitizeAmountInput(_ raw: String) {
        // Keep digits only for whole-dollar gifts via number pad.
        let digits = raw.filter(\.isNumber)
        // Cap length to avoid absurd values.
        amountInput = String(digits.prefix(7))
    }


    var paymentMethodSummaryTitle: String {
        switch selectedMethod {
        case .bankAccount:
            return "Bank · \(last4(bankDetails.accountNumber))"
        case .zelle:
            return "Zelle"
        case .venmo:
            return "Venmo"
        case .paypal:
            return "PayPal"
        case .creditCard:
            let brand = Self.cardBrandLabel(for: creditCardDetails.cardNumber)
            return "\(brand) · \(last4(creditCardDetails.cardNumber))"
        case .giftFund:
            return "Gift Fund"
        }
    }

    var paymentMethodSummarySubtitle: String {
        switch selectedMethod {
        case .creditCard:
            return "Expires \(creditCardDetails.expiry)"
        case .bankAccount:
            return "ACH · 1 – 3 Business days"
        case .zelle, .venmo, .paypal:
            return "Instant · Free"
        case .giftFund:
            return "Gift balance"
        }
    }

    var paymentDeliveryLabel: String {
        switch selectedMethod {
        case .bankAccount: "1 – 3 Business days"
        case .zelle, .venmo, .paypal, .creditCard, .giftFund: "Instant"
        }
    }

    var paymentFeeLabel: String { "Free" }

    var primaryLogoAsset: String? {
        switch selectedMethod {
        case .zelle: "payZelle"
        case .venmo: "payVenmo"
        case .paypal: "payPayPal"
        case .creditCard: "payVisa"
        default: nil
        }
    }

    var receivedDateLabel: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMMM d, yyyy"
        return formatter.string(from: receivedDate)
    }

    static func formatCurrency(_ value: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        formatter.maximumFractionDigits = 0
        formatter.minimumFractionDigits = 0
        return formatter.string(from: NSNumber(value: value)) ?? "$\(Int(value))"
    }

    private func last4(_ digits: String) -> String {
        let cleaned = digits.filter(\.isNumber)
        guard cleaned.count >= 4 else { return cleaned.isEmpty ? "----" : cleaned }
        return String(cleaned.suffix(4))
    }

    private static func cardBrandLabel(for number: String) -> String {
        let cleaned = number.filter(\.isNumber)
        if cleaned.hasPrefix("4") { return "Visa" }
        if cleaned.hasPrefix("5") { return "Mastercard" }
        return "Card"
    }
}

File: android/app/src/main/java/com/embelife/app/model/GiftModels.kt
package com.embelife.app.model

import java.time.LocalDate

enum class GiftExperienceRoute {
    Confirm,
    PaymentMethod,
    Sent,
    SignUp,
    Received,
}

/** Draft state for the Gift $ Amount flow. */
class GiftDraft {
    var amountInput: String = "1000"
    var selectedMethod: PaymentMethodKind = PaymentMethodKind.CreditCard
    var bankDetails: BankAccountDetails = BankAccountDetails.sample
    var zelleDetails: ContactPaymentDetails = ContactPaymentDetails.zelleSample
    var venmoDetails: ContactPaymentDetails = ContactPaymentDetails.venmoSample
    var paypalDetails: ContactPaymentDetails = ContactPaymentDetails.paypalSample
    var creditCardDetails: CreditCardDetails = CreditCardDetails.sample
    var recipientLabel: String = "Gift money to EmBeLife User"
    var senderDisplayName: String = "J. R"
    var receivedDate: LocalDate = LocalDate.of(2026, 5, 27)

    var amount: Double
        get() = amountInput.filter { it.isDigit() || it == '.' }.toDoubleOrNull() ?: 0.0
        set(value) {
            val clamped = value.coerceAtLeast(0.0)
            amountInput = if (clamped == kotlin.math.floor(clamped)) {
                clamped.toInt().toString()
            } else {
                "%.2f".format(clamped)
            }
        }

    val isValidGiftAmount: Boolean get() = amount > 0
    val amountLabel: String get() = formatCurrency(amount)

    fun applyPreset(preset: Double) {
        amount = preset
    }

    fun sanitizeAmountInput(raw: String) {
        amountInput = raw.filter { it.isDigit() }.take(7)
    }

    val paymentMethodSummaryTitle: String
        get() = when (selectedMethod) {
            PaymentMethodKind.BankAccount -> "Bank · ${last4(bankDetails.accountNumber)}"
            PaymentMethodKind.Zelle -> "Zelle"
            PaymentMethodKind.Venmo -> "Venmo"
            PaymentMethodKind.Paypal -> "PayPal"
            PaymentMethodKind.CreditCard -> "${cardBrandLabel(creditCardDetails.cardNumber)} · ${last4(creditCardDetails.cardNumber)}"
            PaymentMethodKind.GiftFund -> "Gift Fund"
        }

    val paymentMethodSummarySubtitle: String
        get() = when (selectedMethod) {
            PaymentMethodKind.CreditCard -> "Expires ${creditCardDetails.expiry}"
            PaymentMethodKind.BankAccount -> "ACH · 1 – 3 Business days"
            PaymentMethodKind.Zelle, PaymentMethodKind.Venmo, PaymentMethodKind.Paypal -> "Instant · Free"
            PaymentMethodKind.GiftFund -> "Gift balance"
        }

    val paymentDeliveryLabel: String
        get() = when (selectedMethod) {
            PaymentMethodKind.BankAccount -> "1 – 3 Business days"
            else -> "Instant"
        }

    companion object {
        val amountPresets = listOf(50.0, 100.0, 200.0, 500.0, 1_000.0, 2_000.0)
        val giftPaymentMethods = listOf(
            PaymentMethodKind.BankAccount,
            PaymentMethodKind.Zelle,
            PaymentMethodKind.Venmo,
            PaymentMethodKind.Paypal,
            PaymentMethodKind.CreditCard,
        )

        fun formatCurrency(value: Double): String = "$${value.toInt()}"

        private fun last4(digits: String): String {
            val cleaned = digits.filter { it.isDigit() }
            return when {
                cleaned.length >= 4 -> cleaned.takeLast(4)
                cleaned.isEmpty() -> "----"
                else -> cleaned
            }
        }

        private fun cardBrandLabel(number: String): String {
            val cleaned = number.filter { it.isDigit() }
            return when {
                cleaned.startsWith("4") -> "Visa"
                cleaned.startsWith("5") -> "Mastercard"
                else -> "Card"
            }
        }
    }
}

enum class SendGiftAction(val title: String) {
    GiftAmount("Gift $ Amount"),
    GiftService("Gift Service"),
    ReceiveGift("Receive Gift"),
}

enum class NotesScreenPhase { Welcome, Ready, Conversation }

enum class VoiceConversationPhase { Listening, UserTranscript, Playback, AiReply }

data class VoiceNoteMessage(
    val id: String,
    val role: Role,
    val text: String,
    val timeLabel: String,
    val showsWaveform: Boolean = false,
) {
    enum class Role { User, Assistant }
}

object VoiceNoteSample {
    const val userTranscript = "hi, im looking for child care for my kid..."
    const val assistantReply = "I got it, let me look for some options for you"
    const val userTime = "08:33"
    const val audioDuration = "0:08"
}
